storage/src/main/java/tech/pegasys/teku/storage/Store.java

```diff
@@ -403,6 +561,74 @@ public VoteTracker getVote(UnsignedLong validatorIndex) {
               });
     }
 
+    private Map<Bytes32, SignedBlockAndState> getFinalizedChainData(
+        final SignedBlockAndState previouslyFinalizedBlock,
+        final SignedBlockAndState newlyFinalizedBlock) {
+      final Map<Bytes32, SignedBlockAndState> finalizedChainData = new HashMap<>();
+
+      SignedBlockAndState oldestFinalizedBlock = newlyFinalizedBlock;
+      SignedBlockAndState currentBlock = newlyFinalizedBlock;
+      while (currentBlock != null
+          && currentBlock.getSlot().compareTo(previouslyFinalizedBlock.getSlot()) > 0) {
+        finalizedChainData.put(currentBlock.getRoot(), currentBlock);
+        oldestFinalizedBlock = currentBlock;
+        currentBlock = getBlockAndState(currentBlock.getParentRoot()).orElse(null);
+      }
+
+      // Make sure we capture all finalized blocks
+      if (!oldestFinalizedBlock.getParentRoot().equals(previouslyFinalizedBlock.getRoot())) {
+        throw new IllegalStateException("Unable to retrieve all finalized blocks");
+      }
+
+      return finalizedChainData;
+    }
+
+    private Map<UnsignedLong, Set<Bytes32>> getPrunedHotBlockRoots(
+        final Checkpoint newFinalizedCheckpoint,
+        final SignedBeaconBlock finalizedBlock,
+        final Map<Bytes32, SignedBeaconBlock> newBlocks) {
+      final UnsignedLong finalizedSlot = newFinalizedCheckpoint.getEpochStartSlot();
+      Map<UnsignedLong, Set<Bytes32>> prunedBlockRoots = new HashMap<>();
+
+      // Build combined index from slot to block root for new and existing blocks
+      final NavigableMap<UnsignedLong, Set<Bytes32>> slotToHotBlockRootIndex = new TreeMap<>();
+      indexBlockRootsBySlot(slotToHotBlockRootIndex, newBlocks.values());
+      slotToHotBlockRootIndex.putAll(rootsBySlotLookup);
```

Comment:
Don't these need to be the other way around?  First put all the existing data into the map, then add the new values? Otherwise if the new blocks are from the same slot as any existing block the `putAll` will overwrite the set (whereas it should merge them).

Reply:
yes - good catch!  will make sure we have some tests around this.